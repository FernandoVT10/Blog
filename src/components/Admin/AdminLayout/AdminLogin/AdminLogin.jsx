import ApiController from "../../../../services/ApiController";

import { useState } from "react";

import "./AdminLogin.scss";

export default ({ setIsAuthenticated }) => {
    const [username, setUsername] = useState("");
    const [password, setPassword] = useState("");
    const [errorMessage, setErrorMessage] = useState("");
    const [loading, setLoading] = useState(false);

    const handleForm = e => {
        e.preventDefault();

        setErrorMessage("");
        setLoading(true);

        ApiController.post("users/login/", { username, password })
        .then(res => {
            if(res.data) {
                window.localStorage.setItem("token", res.data.token);

                setIsAuthenticated(true);
            } else {
                setErrorMessage(res.errors[0].message);
            }

            setLoading(false);
        });
    };

    return (
        <div className="admin-login">
            <div className="admin-login__form-container">
                <form onSubmit={handleForm}>
                    <h2 className="admin-login__title">Contact Me</h2>

                    <input
                    type="text"
                    className="formulary__input admin-login__input"
                    placeholder="Username"
                    autoComplete="username"
                    onChange={({ target: { value } }) => setUsername(value)}
                    required />

                    <input
                    type="password"
                    className="formulary__input admin-login__input"
                    placeholder="Password"
                    autoComplete="password"
                    onChange={({ target: { value } }) => setPassword(value)}
                    required />

                    { errorMessage &&
                        <p className="formulary__message formulary__message--error">
                            <i className="fas fa-times-circle mr-2" aria-hidden="true"></i>
                            { errorMessage }
                        </p>
                    }

                    { loading ?
                        <div className="admin-login__loader">
                            <span className="spinner-border"></span>
                        </div>
                        :
                        <div className="admin-login__submit-button-container">
                            <button className="formulary__submit-button text-center">
                                Login
                            </button>
                        </div>
                    }
                </form>
            </div>
        </div>
    );
};