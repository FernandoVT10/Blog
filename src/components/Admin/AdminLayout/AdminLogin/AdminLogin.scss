.admin-login {
    display: flex;
    justify-content: center;
    align-items: center;
    min-height: 100vh;

    &__form-container {
        padding: 20px;
        background: var(--primary);
        max-width: 500px;
    }

    &__title {
        color: var(--white);
        font-size: 1.5em;
        font-weight: bold;
    }

    &__input {
        margin: 10px 0;
    }

    &__submit-button-container {
        text-align: center;
        margin: 10px;
    }

    &__loader {
        margin: 10px;
        text-align: center;
        color: var(--secondary);
    }
}