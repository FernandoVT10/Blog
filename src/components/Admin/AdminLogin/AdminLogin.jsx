import Layout from "../../../components/Layout";
import { useState } from "react";

import "./AdminLogin.scss";
import Api from "../../../ApiController";

export default () => {
    const [username, setUsername] = useState("");
    const [password, setPassword] = useState("");
    const [errorMessage, setErrorMessage] = useState("");

    const login = e => {
        e.preventDefault();

        Api.post("users/login/", { username, password })
        .then(data => {
            if(data.status) {
                window.localStorage.setItem("token", data.token);

                window.location.reload();
            } else {
                setErrorMessage(data.error.message);
            }
        });
    };

    return (
        <Layout>
            <div className="container-fluid">
                <div className="row justify-content-center">
                    <div className="col-12 col-md-6">
                        <form onSubmit={login}>
                            <h2 className="title mb-3">Contact Me</h2>

                            <input
                            type="text"
                            className="formulary__input"
                            placeholder="Username"
                            autoComplete="username"
                            onChange={({ target: { value } }) => setUsername(value)}
                            required />

                            <input
                            type="password"
                            className="formulary__input"
                            placeholder="Password"
                            autoComplete="password"
                            onChange={({ target: { value } }) => setPassword(value)}
                            required />

                            <p className="text-danger">{ errorMessage }</p>

                            <button className="submit-button">
                                Login
                            </button>
                        </form>
                    </div>
                </div>
            </div>
        </Layout>
    );
};