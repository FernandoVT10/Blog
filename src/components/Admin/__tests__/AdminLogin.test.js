import { act, Simulate } from "react-dom/test-utils";
import { render } from "react-dom";
import AdminLogin from "../AdminLogin/";

let container;

describe('<AdminLogin/> Component', () => {
    beforeEach(() => {
        fetchMock.doMock();

        container = document.createElement('div');
        document.body.appendChild(container);
    });
    
    afterEach(() => {
        document.body.removeChild(container);
        container = null;
    });

    it("It should display an error message", async () => {
        fetch.mockResponse(JSON.stringify(
            {
                status: false,
                error: { message: "Test error" }
            }
        ));

        render(<AdminLogin/>, container);

        const form = container.querySelector("form");
        const inputs = container.querySelectorAll(".formulary__input");

        act(() => {
            inputs.forEach((input) => {
                input.value = "Test value";
                Simulate.change(input);
            });
        });

        await act(async () => {
            Simulate.submit(form);
        });

        const p = container.querySelector("p.text-danger");

        expect(p.textContent).toBe("Test error");
    });

    it("It should set the token on local storage and reload the page", async () => {
        const reloadMock = jest.fn();

        Object.defineProperty(window, "location", {
            value: {
                reload: reloadMock
            }
        });

        fetch.mockResponse(JSON.stringify(
            {
                status: true,
                token: "Mock token"
            }
        ));

        render(<AdminLogin/>, container);

        const form = container.querySelector("form");
        const inputs = container.querySelectorAll(".formulary__input");

        act(() => {
            inputs.forEach((input) => {
                input.value = "Test value";
                Simulate.change(input);
            });
        });

        await act(async () => {
            Simulate.submit(form);
        });

        expect(window.localStorage.getItem("token")).toBe("Mock token");
        expect(reloadMock).toHaveBeenCalledTimes(1);
    });
});