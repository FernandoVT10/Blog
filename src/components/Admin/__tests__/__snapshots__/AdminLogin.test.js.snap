// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdminLogin/> Component Check if renders correctly 1`] = `
<div>
  <div>
    <div
      class="container-fluid"
    >
      <div
        class="row justify-content-center"
      >
        <div
          class="col-12 col-md-6"
        >
          <form>
            <h2
              class="title mb-3"
            >
              Contact Me
            </h2>
            <input
              autocomplete="username"
              class="formulary__input"
              placeholder="Username"
              required=""
              type="text"
            />
            <input
              autocomplete="password"
              class="formulary__input"
              placeholder="Password"
              required=""
              type="password"
            />
            <p
              class="text-danger"
            />
            <button
              class="submit-button"
            >
              Login
            </button>
          </form>
        </div>
      </div>
    </div>
    <script
      src="https://cdn.jsdelivr.net/npm/chart.js@2.9.3/dist/Chart.min.js"
    />
    <script
      src="https://code.jquery.com/jquery-3.4.1.slim.min.js"
    />
    <script
      src="https://cdn.jsdelivr.net/npm/popper.js@1.16.0/dist/umd/popper.min.js"
    />
    <script
      src="https://stackpath.bootstrapcdn.com/bootstrap/4.4.1/js/bootstrap.min.js"
    />
    <script
      src="https://kit.fontawesome.com/63ef8f1397.js"
    />
  </div>
</div>
`;
